ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Worker.java

```diff
@@ -233,39 +234,36 @@ void gatherStats() {
     private final CompactionTxn compactionTxn;
     private final String tableName;
     private final HiveConf conf;
-    private final long txnTimeout;
+    private final AtomicBoolean shouldLogError;
 
     public CompactionHeartbeater(CompactionTxn compactionTxn, String tableName, HiveConf conf) {
       this.tableName = Objects.requireNonNull(tableName);
       this.compactionTxn = Objects.requireNonNull(compactionTxn);
       this.conf = Objects.requireNonNull(conf);
-
-      this.txnTimeout = MetastoreConf.getTimeVar(conf, MetastoreConf.ConfVars.TXN_TIMEOUT, TimeUnit.MILLISECONDS);
+      this.shouldLogError = new AtomicBoolean(true);
 
       setDaemon(true);
       setPriority(MIN_PRIORITY);
       setName("CompactionHeartbeater-" + compactionTxn.getTxnId());
     }
 
+    public void shouldLogError(boolean shouldLogError) {
+      this.shouldLogError.set(shouldLogError);
+    }
+
     @Override
     public void run() {
       LOG.debug("Heartbeating compaction transaction id {} for table: {}", compactionTxn, tableName);
-
       IMetaStoreClient msc = null;
       try {
         // Create a metastore client for each thread since it is not thread safe
         msc = HiveMetaStoreUtils.getHiveMetastoreClient(conf);
```

Comment:
This way we are getting a HMS client for every heartbeat... maybe we only need one for each instance

Reply:
That makes sense, and I was debating that internally as well. However, it would complicate the cleanup of the thread quite a bit. Since the thread runs infrequently (every 2.5 minutes or so) I think the cost of client creation should be pretty low, and not worth the extra complexity of storing it in a field and worry about clean up in the various scenarios. Let me know if that makes sense.

Reply:
I mean, it's closed in the finally block.... Honestly I'm not sure how expensive client creation is. @pvary do you know?

Reply:
it wouldn't be closed in the finally block if it's cached as an object field. In that case, we'd need the `CompactionHeartbeater` to have a close() method, and invoke it somewhere, for it to close its cached client.

Reply:
HMS client creation itself is not that significant. OTOH if we can save resources we should do so, just make sure that we can recover if there is an error, like HMS restart, etc.